tools/net-common.in

```diff
@@ -63,6 +63,68 @@ function fix_mac {
   ip link set $INTERFACE address $FIXED_MAC || true
 }
 
+function bridge_vlan_add {
+  local VID="${1}"
+  shift
+
+  # when the VID is in use by traditional VLAN interface, then the
+  # tap on the VLAN aware bridge will not receive traffic
+  if [ -r /proc/net/vlan/config ]; then
+    if ! awk -F '[| ]*' -v vlan="^${VID}$" 'match($2, vlan) { exit 1 }' /proc/net/vlan/config; then
+      echo "VLAN ${VID} is in use by interface $(awk -F '[| ]*' -v vlan="^${VID}$" 'match($2, vlan) { print $1 }' /proc/net/vlan/config)"
+      exit 1
+    fi
+  fi
+
+  # add the VLAN to the tap
+  # $@ is either empty or contains "pvid untagged"
+  bridge vlan add dev ${INTERFACE} vid ${VID} "${@}" master
+  # add the VLAN to the bridge uplinks
+  local uplink
+  for uplink in ${BRIDGE_UPLINKS}; do
+    bridge vlan add dev ${uplink} vid ${VID} master
```

Comment:
It would be good to add some logging here, as we are touching interfaces other than the instance's tap and the bridge itself.